pytensor/tensor/c_code/dimshuffle.c

```diff
@@ -1,82 +1,83 @@
-  }
-
-  /*
-    res = res.transpose(self.transposition)
-  */
-  PyArrayObject *transposed_input =
-      (PyArrayObject *)PyArray_Transpose(_input, &permute);
-
-  Py_DECREF(_input);
-
-  PyDimMem_FREE(permute.ptr);
-
-  npy_intp *res_shape = PyArray_DIMS(transposed_input);
-  npy_intp N_shuffle = PyArray_SIZE(params->shuffle);
-  npy_intp N_augment = PyArray_SIZE(params->augment);
-  npy_intp N = N_augment + N_shuffle;
-  npy_intp *_reshape_shape = PyDimMem_NEW(N);
-
-  if (_reshape_shape == NULL) {
-    PyErr_NoMemory();
-    return 1;
-  }
+    if (!PyArray_IS_C_CONTIGUOUS(params->_new_order)) {
+        PyErr_SetString(PyExc_RuntimeError, "DimShuffle: param _new_order must be C-contiguous.");
+        return 1;
+    }
+    new_order = (npy_int64*) PyArray_DATA(params->_new_order);
+    nd_in = (npy_intp)(params->input_ndim);
+    nd_out = PyArray_SIZE(params->_new_order);
 
-  /*
-    shape = list(res.shape[: len(self.shuffle)])
-    for augm in self.augment:
-        shape.insert(augm, 1)
-  */
-  npy_intp aug_idx = 0;
-  int res_idx = 0;
-  for (npy_intp i = 0; i < N; i++) {
-    if (aug_idx < N_augment &&
-        i == *((npy_intp *)PyArray_GetPtr(params->augment, &aug_idx))) {
-      _reshape_shape[i] = 1;
-      aug_idx++;
-    } else {
-      _reshape_shape[i] = res_shape[res_idx];
-      res_idx++;
+    if (PyArray_NDIM(input) != nd_in) {
+        PyErr_SetString(PyExc_NotImplementedError, "DimShuffle: Input has less dimensions than expected.");
+        return 1;
     }
-  }
 
-  PyArray_Dims reshape_shape = {.ptr = _reshape_shape, .len = (int)N};
+    if (*res)
+        Py_XDECREF(*res);
+
+    Py_INCREF((PyObject*)input);
+
+    // Compute new dimensions and strides
+    dimensions = (npy_intp*) malloc(nd_out * sizeof(npy_intp));
+    strides = (npy_intp*) malloc(nd_out * sizeof(npy_intp));
+    if (dimensions == NULL || strides == NULL) {
+        PyErr_NoMemory();
+        free(dimensions);
+        free(strides);
+        return 1;
+    };
+
+    npy_intp original_size = PyArray_SIZE(input);
+    npy_intp new_size = 1;
+    for (npy_intp i = 0; i < nd_out; ++i) {
+        if (new_order[i] != -1) {
+            dimensions[i] = PyArray_DIMS(input)[new_order[i]];
+            strides[i] = PyArray_DIMS(input)[new_order[i]] == 1 ? 0 : PyArray_STRIDES(input)[new_order[i]];
+        } else {
+            dimensions[i] = 1;
+            strides[i] = 0;
+        }
+        new_size *= dimensions[i];
+    }
 
-  /* res = res.reshape(shape) */
-  *res = (PyArrayObject *)PyArray_Newshape(transposed_input, &reshape_shape,
-                                           NPY_CORDER);
+    if (original_size != new_size) {
+        PyErr_SetString(PyExc_ValueError, "DimShuffle: Attempting to squeeze axes with size not equal to one.");
+        free(dimensions);
+        free(strides);
+        return 1;
+    }
 
-  Py_DECREF(transposed_input);
+    // Create the new array.
+    *res = (PyArrayObject*)PyArray_New(&PyArray_Type, nd_out, dimensions,
```

Comment:
I'm not really used to writing python C api code, so I could be very wrong, but shouldn't the decref of *res be just before this here? Otherwise we might be dropping an array that the caller still want's to use?
And the incref on input should only happen after we check that the new array creation was successful?

Reply:
The decref was always like this, I think PyTensor preallocates the output (with None or the last call if gc is turned off) so if you don't need it you can just decref it immediately?

Reply:
You were saying something els... We don't use the original res anywhere? Why does it matter when we decref?

Reply:
Because we have some checks between the call to decref and when we change the output pointer.
And in those checks we might exit early and never change the output pointer.

Reply:
So, I *think* we should either:
- decref the thing pointed to by the output pointer and then change the output pointer to something else that we either increfed or created ourselves and decref
- Write our results into the thing pointed to by the output pointer.
Mixing them doesn't sound like a good idea...

Reply:
But we never care for the original output even if we exit early no?

Reply:
I'll check again but I be surprised if this was problematic as it was always like this since early Theano

Reply:
I couldn't find the code that calls this function yet, so I don't know for sure what happens. But I'd expect something that usually something like the following should happen:

```
generate the output array (or incref it if it already exists)
Call the function that does all the work
Check the return code. If the function failed, decref the outputs.
```

Now, if we return a failure code but decref the outputs ourselves, that'd be a double free, and could potentially segfault.
I guess it's also possible that it doesn't decref, but I think then we'd have a memory leak in those cases where we return an error but don't call decref ourselves.

Reply:
Okay so are you saying I should move the decref/incref after the new size check?